File: Java/Medium/RuntimeError/minimum-fuel-cost-to-report-to-the-capital_7.java
class Solution {
    public long minimumFuelCost(int[][] roads, int seats) {
        
        int n = roads.length + 1;
        List<Integer>[] adj = new List[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        for (int[] road : roads) {
            adj[road[0]].add(road[1]);
            adj[road[1]].add(road[0]);
        }

        int[] sub = new int[n];
        dfs(adj, sub, 0, -1);

        long[] dp = new long[n];
        Arrays.fill(dp, Long.MAX_VALUE);
        dp[0] = 0;
        dfs(adj, dp, sub, 0, -1, seats);
        return dp[0];
    }

    private void dfs(List<Integer>[] adj, int[] sub, int u, int p) {
        sub[u] = 1;
        for (int v : adj[u]) {
            if (v == p) {
                continue;
            }
            dfs(adj, sub, v, u);
            sub[u] += sub[v];
        }
    }

    private void dfs(List<Integer>[] adj, long[] dp, int[] sub, int u, int p, int seats) {
        for (int v : adj[u]) {
            if (v == p) {
                continue;
            }
            dfs(adj, dp, sub, v, u, seats);
            dp[u] += dp[v] + sub[v];
        }

        long min = Long.MAX_VALUE;
        for (int v : adj[u]) {
            if (v == p) {
                continue;
            }
            min = Math.min(min, dp[u] - dp[v] - sub[v] + dp[v] + (sub[v] + seats - 1) / seats);
        }
        dp[u] = Math.min(dp[u], min);
    }
}

File: C++/Medium/WrongAnswer/minimum-fuel-cost-to-report-to-the-capital_7.cpp
class Solution {
public:
    long long minimumFuelCost(vector<vector<int>>& roads, int seats) {
        
        int n = roads.size() + 1;
        vector<vector<int>> g(n, vector<int>());
        for (auto& e: roads) {
            g[e[0]].push_back(e[1]);
            g[e[1]].push_back(e[0]);
        }
        vector<int> sub(n, 1);
        vector<int> f(n, 0);
        vector<int> dp(n, 0);
        vector<int> g1(n, 0);
        vector<int> g2(n, 0);
        vector<int> g3(n, 0);
        function<void(int, int)> dfs = [&](int u, int p) {
            for (int v: g[u]) {
                if (v != p) {
                    dfs(v, u);
                    sub[u] += sub[v];
                    f[u] += f[v] + sub[v];
                }
            }
        };
        dfs(0, -1);
        function<void(int, int, int)> dfs2 = [&](int u, int p, int s) {
            dp[u] = s;
            for (int v: g[u]) {
                if (v != p) {
                    dfs2(v, u, s + f[0] - f[v] - sub[v] + n - sub[v]);
                }
            }
        };
        dfs2(0, -1, 0);
        function<void(int, int)> dfs3 = [&](int u, int p) {
            for (int v: g[u]) {
                if (v != p) {
                    g1[v] = g1[u] + f[u] - f[v] - sub[v] + n - sub[v];
                    g2[v] = min(g2[u], g1[v]);
                    g3[v] = min(g3[u], g1[v]);
                    dfs3(v, u);
                }
            }
        };
        dfs3(0, -1);
        long long ans = 0;
        for (int i = 0; i < n; ++i) {
            ans += min({dp[i], g1[i], g2[i], g3[i]});
        }
        return ans / 2;
    }
};